core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -405,6 +415,13 @@ private[spark] class Executor(
           setTaskFinishedAndClearInterruptStatus()
           execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
 
+        case t: Throwable if task.context.fetchFailed.isDefined =>
+          // tbere was a fetch failure in the task, but some user code wrapped that exception
+          // and threw something else.  Regardless, we treat it as a fetch failure.
+          val reason = task.context.fetchFailed.get.toTaskFailedReason
+          setTaskFinishedAndClearInterruptStatus()
+          execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
```

Comment:
do you mean the msg I added about "TID ${taskId} completed successfully though internally it encountered unrecoverable fetch failures!"?  I wouldn't think we'd want to log anything special here.  I'm trying to make this a "normal" code path.  The user is allowed to allowed to do this.  (sparksql already does.)

we could log a warning, but then this change should be accompanied by auditing the code and making sure we never do this ourselves.